pyccl/nl_pt/ept.py

```diff
@@ -0,0 +1,700 @@
+            self.ia_tt = self.pt.IA_tt(
+                pklz0, P_window=self.fastpt_par['P_window'],
+                C_window=self.fastpt_par['C_window'])
+            self.ia_mix = self.pt.IA_mix(
+                pklz0, P_window=self.fastpt_par['P_window'],
+                C_window=self.fastpt_par['C_window'])
+
+        # b1/bk power spectrum
+        pks = {}
+        if 'nonlinear' in [self.b1_pk_kind, self.bk2_pk_kind]:
+            pks['nonlinear'] = np.array([cosmo.nonlin_matter_power(self.k_s, a)
+                                         for a in self.a_s])
+        if 'linear' in [self.b1_pk_kind, self.bk2_pk_kind]:
+            pks['linear'] = np.array([cosmo.linear_matter_power(self.k_s, a)
+                                      for a in self.a_s])
+        if 'pt' in [self.b1_pk_kind, self.bk2_pk_kind]:
+            if 'linear' in pks:
+                pk = pks['linear']
+            else:
+                pk = np.array([cosmo.linear_matter_power(self.k_s, a)
+                               for a in self.a_s])
+            # Add SPT correction
+            pk += self._g4[:, None] * self.one_loop_dd[0][None, :]
+            pks['pt'] = pk
+        self.pk_b1 = pks[self.b1_pk_kind]
+        self.pk_bk = pks[self.bk2_pk_kind]
+
+        # Reset template power spectra
+        self._pk2d_temp = {}
+        self._pt_init = True
+
+    def _get_pgg(self, tr1, tr2):
+        """ Get the number counts auto-spectrum at the internal
+        set of wavenumbers and scale factors.
+
+        Args:
+            tr1 (:class:`~pyccl.nl_pt.tracers.PTTracer`): first
+                tracer to correlate.
+            tr2 (:class:`~pyccl.nl_pt.tracers.PTTracer`): first
+                tracer to correlate.
+
+        Returns:
+            array_like: 2D array of shape `(N_a, N_k)`, where `N_k` \
+                is the size of this object's `k_s` attribute, and \
+                `N_a` is the size of the object's `a_s` attribute.
+        """
+        self._check_pt_init()
+        # Get Pk templates
+        Pd1d1 = self.pk_b1
+        Pd1d2 = self._g4[:, None] * self.dd_bias[2][None, :]
+        Pd2d2 = self._g4[:, None] * self.dd_bias[3][None, :]
+        Pd1s2 = self._g4[:, None] * self.dd_bias[4][None, :]
+        Pd2s2 = self._g4[:, None] * self.dd_bias[5][None, :]
+        Ps2s2 = self._g4[:, None] * self.dd_bias[6][None, :]
+        Pd1p3 = self._g4[:, None] * self.dd_bias[8][None, :]
+        Pd1k2 = self.pk_bk * (self.k_s**2)[None, :]
+
+        # Get biases
+        b11 = tr1.b1(self.z_s)
+        b21 = tr1.b2(self.z_s)
+        bs1 = tr1.bs(self.z_s)
+        bk21 = tr1.bk2(self.z_s)
+        b3nl1 = tr1.b3nl(self.z_s)
+        b12 = tr2.b1(self.z_s)
+        b22 = tr2.b2(self.z_s)
+        bs2 = tr2.bs(self.z_s)
+        bk22 = tr2.bk2(self.z_s)
+        b3nl2 = tr2.b3nl(self.z_s)
+
+        s4 = 0.
+        if self.fastpt_par['sub_lowk']:
+            s4 = self._g4 * self.dd_bias[7]
+            s4 = s4[:, None]
```

Comment:
If sub_lok is True, this assigns 0 to s4 first, and then reassigns it with another value. The proper way to do that is to use an if-else statement. This is going to be a waste of space though, that's why I prefer the one-liners:
```python
        s4 = ((self._g4 * self.dd_bias[7])[:, None]
              if self.fastpt_par['sub_lowk']
              else 0.)
```

Reply:
This seems fine to me as it is.

Reply:
It is inefficient because if `sub_lok` is `True` memory will be allocated twice. If we're going to be strict about efficiency like in `halos/` we have to do it everywhere, right? xD
I don't care about that to be fair, I'm just pointing it out, you can leave it as it is.